Parallel

farhadrclass opened the pull request:
### Summary 
The function bmark_solvers_parallel is designed to benchmark a set of solvers on a set of problems in parallel, leveraging multi-core CPUs for performance improvements on large sets of problems or computationally intensive solvers. By running in parallel, it aims to significantly reduce runtime.

### Key Functionality
Parallel Execution: Running solvers concurrently across multiple threads to enhance benchmarking speed.
CPU Utilization: This function is optimized for multi-core CPUs, and users should configure the number of threads based on their machine’s capabilities. If not specified, it defaults to single-threaded execution, similar to bmark_solver().
To specify the number of threads, users can set the JULIA_NUM_THREADS environment variable before starting Julia:

```
export JULIA_NUM_THREADS=4  # for 4 threads
```
Then, within the session, verify the number of threads with:

```

Threads.nthreads()
```
### Questions and Considerations
The following are my concerns and Questions:
 
**Examples for Documentation:**
Examples should include a comparison of single-threaded versus multi-threaded runs, showing how to use bmark_solvers_parallel with different thread configurations.
Consider an example with varying numbers of solvers and problems to illustrate performance benefits.

**Combining bmark_solvers_parallel and bmark_solvers:**
Combining both functions could streamline the API. Adding a threads=4 default argument (with threads=1 for sequential execution) would allow users to control threading from within the function, reducing the need for environment variable configuration and simplifying usage.

**Parallelizing Problems vs. Solvers:**
For now, parallelizing the solvers is a good first step. In the future, should we consider parallelizing over both solvers and problems to maximize performance on systems with many cores.

**Multi-CPU or Single Multi-Core CPU:**
This implementation is best suited to single CPUs with multiple cores (e.g., 4-8 cores typical for desktop CPUs). Supporting multi-CPU systems would require additional architecture-specific optimizations and may be better handled as a future extension if user demand and workload complexity increase.

dpo commented on src/bmark_solvers.jl:
```diff
@@ -23,3 +26,44 @@ function bmark_solvers(solvers::Dict{Symbol, <:Any}, args...; kwargs...)
+  solver_keys = collect(keys(solvers))
+  solver_values = collect(values(solvers))
+
+  Threads.@threads for i in eachindex(solver_keys)
```
Why not simply add threads to the existing `bmark_solvers()` function?

MaxenceGollier commented:
Thank you @farhadrclass, this will be very helpful for me. 
As already mentionned by @dpo, wouldn't it be easier to just modify `bmark_solvers` ? 
I had this in mind:
Modify 
https://github.com/JuliaSmoothOptimizers/SolverBenchmark.jl/blob/360eb42a04beca2a57ee6e894481ebe60695a6b0/src/bmark_solvers.jl#L20
with
```
Threads.@threads for (name, solver) in solvers
```
We could also modify 
https://github.com/JuliaSmoothOptimizers/SolverBenchmark.jl/blob/360eb42a04beca2a57ee6e894481ebe60695a6b0/src/run_solver.jl#L96
with 
```
Threads.@threads for (id, problem) in enumerate(problems) 
```
although the second option might require a little more care on `stats` because each thread pushes into it, it is maybe preferrable because we would still have each solver printed out sequentially.
In all cases, thank you again, I hope this will be merged soon !

farhadrclass commented:
@MaxenceGollier 
thanks for your feedback, yes the idea was t include them both in there,
I would suggest the solvers, (vs. Problems vs. both) the issue in the problems is that we have many of them and once we update it then writing to stats would be problematic, and for time it is actually the same if we do problems or solvers, doing both might be more challenging since we need to coordinate when each is finish and how we write to the stats.,
I will push the code soon for review

farhadrclass commented:
@dpo  the checks keep failing and says can not find cons_nln method, not sure why

dpo commented:
I don't see that error. Where is it?

You can drop Julia 1.6 and replace it with 1.10.

dpo commented on src/bmark_solvers.jl:
```diff
@@ -15,11 +17,12 @@ Any keyword argument accepted by `solve_problems`
-    stats[name] = solve_problems(solver, name, args...; kwargs...)
+  @threads for (name, solver) in collect(pairs(solvers))
+      @info "running solver $name"
+      stats[name] = solve_problems(solver, name, args...; kwargs...)
```
Indentation is 2 spaces.

dpo commented on src/bmark_solvers.jl:
```diff
@@ -1,9 +1,11 @@
 
-Run a set of solvers on a set of problems.
+Run a set of solvers on a set of problems. If Threads.@threads is larger than 1, the solvers will be run in parallel.
+Note that the @threads macro requires that the number of threads is set before the Julia process starts, using the -t or --threads command-line argument, or the JULIA_NUM_THREADS environment variable.
```
```suggestion
Note that the @threads macro requires that the number of threads be set before the Julia process starts, using the -t or --threads command-line argument, or the JULIA_NUM_THREADS environment variable.
```

farhadrclass commented:
> I don't see that error. Where is it?
> 
> You can drop Julia 1.6 and replace it with 1.10.

The issue is the night build

farhadrclass commented:
> I don't see that error. Where is it?
> 
> You can drop Julia 1.6 and replace it with 1.10.

do you want me to do it in this PR or create a new one

tmigot commented:
Hi guys! I tried to explain here #172 why the Examples fails.

MaxenceGollier commented:
@farhadrclass
As I was expecting, this seems to fail when benchmarking with CUTEst:
```
using SolverCore
using CUTEst

import SolverCore.dummy_solver

function test_cutest()
  problem_names = CUTEst.select(min_con=1, max_con = 3, max_var= 3, only_equ_con=true, only_free_var=true)[1:5]
  problem_list = (CUTEstModel(name) for name in problem_names)
  solvers = Dict(:dummy_1 => dummy_solver, :dummy_2 => dummy_solver)
  stats = bmark_solvers(solvers, problem_list)
end

test_cutest()
```
produces exceptions 
```
[ Info: running solver dummy_2
[ Info: running solver dummy_1
At line 398 of file ../src/decode/sifdecoder_standalone.f90 (unit = 57)
Fortran runtime error: File cannot be deleted

Error termination. Backtrace:

Could not print backtrace: libbacktrace could not find executable to open
#0  0x8562adda
#1  0x855b0d51
#2  0x854d47db
#3  0x85597743
#4  0x625310ff
#5  0x4016e4
#6  0x40310f
#7  0x4012f2
#8  0x401405
#9  0xc88e259c
#10  0xc8aeaf37
#11  0xffffffff

gfortran.exe: error: ELFUN.o: No such file or directory
ERROR: LoadError: TaskFailedException

    nested task error: failed process: Process(`'C:\Users\mgoll\.julia\artifacts\fdff308295487f361ef6e8dc2d27f5abe8a6eee9\mingw64\bin\gfortran.exe' -shared -o libWAYSEA1NE_double.dll ELFUN.o -Wl,--whole-archive 'C:\Users\mgoll\.julia\artifacts\eacd3c5d669a4b2c1bd3b7aad671b55cd097cd99\lib\libcutest_double.a' -Wl,--no-whole-archive`, ProcessExited(1)) [1]

    Stacktrace:
      [1] pipeline_error
        @ .\process.jl:598 [inlined]
      [2] run(::Cmd; wait::Bool)
        @ Base .\process.jl:513
      [3] run
        @ .\process.jl:510 [inlined]
      [4] (::CUTEst.var"#9#10"{String})()
        @ CUTEst C:\Users\mgoll\.julia\packages\CUTEst\Eiaqh\src\sifdecoder.jl:218
      [5] cd(f::CUTEst.var"#9#10"{String}, dir::String)
        @ Base.Filesystem .\file.jl:101
      [6] build_libsif(name::String; precision::Symbol, libsif_folder::String)
        @ CUTEst C:\Users\mgoll\.julia\packages\CUTEst\Eiaqh\src\sifdecoder.jl:175
      [7] (::CUTEst.var"#3#4"{Float64, Bool, Bool, Tuple{}, Base.RefValue{Int32}, Base.RefValue{Int32}, String, String, Symbol})()
        @ CUTEst C:\Users\mgoll\.julia\packages\CUTEst\Eiaqh\src\model.jl:131
      [8] cd(f::CUTEst.var"#3#4"{Float64, Bool, Bool, Tuple{}, Base.RefValue{Int32}, Base.RefValue{Int32}, String, String, Symbol}, dir::String)
        @ Base.Filesystem .\file.jl:101
      [9] CUTEstModel{Float64}(::String; decode::Bool, verbose::Bool, efirst::Bool, lfirst::Bool, lvfirst::Bool)
        @ CUTEst C:\Users\mgoll\.julia\packages\CUTEst\Eiaqh\src\model.jl:124
     [10] CUTEstModel(::String; precision::Symbol, decode::Bool, verbose::Bool, efirst::Bool, lfirst::Bool, lvfirst::Bool)
        @ CUTEst C:\Users\mgoll\.julia\packages\CUTEst\Eiaqh\src\model.jl:93
     [11] CUTEstModel
        @ C:\Users\mgoll\.julia\packages\CUTEst\Eiaqh\src\model.jl:74 [inlined]
     [12] #1
        @ .\none:0 [inlined]
     [13] iterate
        @ .\generator.jl:48 [inlined]
     [14] iterate
        @ .\iterators.jl:206 [inlined]
     [15] iterate
        @ .\iterators.jl:205 [inlined]
     [16] solve_problems(solver::typeof(dummy_solver), solver_name::Symbol, problems::Base.Generator{Vector{String}, var"#1#2"}; solver_logger::Base.CoreLogging.NullLogger, reset_problem::Bool, skipif::SolverBenchmark.var"#39#47", colstats::Vector{Symbol}, info_hdr_override::Dict{Symbol, String}, prune::Bool, kwargs::@Kwargs{})
        @ SolverBenchmark C:\Users\mgoll\SolverBenchmark.jl\src\run_solver.jl:96
     [17] solve_problems(solver::Function, solver_name::Symbol, problems::Base.Generator{Vector{String}, var"#1#2"})
        @ SolverBenchmark C:\Users\mgoll\SolverBenchmark.jl\src\run_solver.jl:30
     [18] macro expansion
        @ C:\Users\mgoll\SolverBenchmark.jl\src\bmark_solvers.jl:24 [inlined]
     [19] (::SolverBenchmark.var"#110#threadsfor_fun#51"{SolverBenchmark.var"#110#threadsfor_fun#50#52"{@Kwargs{}, Tuple{Base.Generator{Vector{String}, var"#1#2"}}, Dict{Symbol, DataFrame}, Vector{Pair{Symbol, typeof(dummy_solver)}}}})(tid::Int64; onethread::Bool)
        @ SolverBenchmark .\threadingconstructs.jl:252
     [20] #110#threadsfor_fun
        @ .\threadingconstructs.jl:219 [inlined]
     [21] (::Base.Threads.var"#1#2"{SolverBenchmark.var"#110#threadsfor_fun#51"{SolverBenchmark.var"#110#threadsfor_fun#50#52"{@Kwargs{}, Tuple{Base.Generator{Vector{String}, var"#1#2"}}, Dict{Symbol, DataFrame}, Vector{Pair{Symbol, typeof(dummy_solver)}}}}, Int64})()
        @ Base.Threads .\threadingconstructs.jl:154

...and 1 more exception.

Stacktrace:
  [1] threading_run(fun::SolverBenchmark.var"#110#threadsfor_fun#51"{SolverBenchmark.var"#110#threadsfor_fun#50#52"{@Kwargs{}, Tuple{Base.Generator{Vector{String}, var"#1#2"}}, Dict{Symbol, DataFrame}, Vector{Pair{Symbol, typeof(dummy_solver)}}}}, static::Bool)
    @ Base.Threads .\threadingconstructs.jl:172
  [2] macro expansion
    @ .\threadingconstructs.jl:189 [inlined]
  [3] #bmark_solvers#49
    @ C:\Users\mgoll\SolverBenchmark.jl\src\bmark_solvers.jl:22 [inlined]
  [4] bmark_solvers
    @ C:\Users\mgoll\SolverBenchmark.jl\src\bmark_solvers.jl:20 [inlined]
  [5] test_cutest()
    @ Main C:\Users\mgoll\SolverBenchmark.jl\test\test_cutest.jl:10
  [6] top-level scope
    @ C:\Users\mgoll\SolverBenchmark.jl\test\test_cutest.jl:13
  [7] include(fname::String)
    @ Main .\sysimg.jl:38
  [8] top-level scope
    @ C:\Users\mgoll\SolverBenchmark.jl\test\runtests.jl:25
  [9] include(fname::String)
    @ Main .\sysimg.jl:38
 [10] top-level scope
    @ none:6
in expression starting at C:\Users\mgoll\SolverBenchmark.jl\test\test_cutest.jl:13
in expression starting at C:\Users\mgoll\SolverBenchmark.jl\test\runtests.jl:25
ERROR: Package SolverBenchmark errored during testing
```
I am trying to investigate what is going on here. Could you try to see if the error reproduces on your machine ? 
for info,
```
julia> versioninfo()
Julia Version 1.11.1
Commit 8f5b7ca12a (2024-10-16 10:53 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 12 × Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, skylake)
Threads: 4 default, 0 interactive, 2 GC (on 12 virtual cores)
```

farhadrclass commented:
@MaxenceGollier  I ran the test on my machine, it passed, however some of the solvers were skipped due to some catch and expect logic, something along the line of the Hessian is not available.
Let me know if you want anything on my side, I would have some time to look at it more on the weekend

MaxenceGollier commented:
> however some of the solvers were skipped due to some catch and expect logic, something along the line of the Hessian is not available.
Yes all dummy solvers will fail on cutest models this is normal.

>@MaxenceGollier I ran the test on my machine, it passed
This is weird my initial guess whas that CUTEst had issued with different models being used at once (i.e decoding multiple sifs at once) I guess this is not the case if it passes on your machine. Could you send me your versioninfo ?

farhadrclass commented:
I rebased on the new update and also added a flag 
```julia 
`threads_enable`: A boolean indicating whether to enable multithreading. Default is `false`.
```

farhadrclass commented:
@dpo  I updated the code and run all the test both in ENV["JULIA_NUM_THREADS"] = "1" and ENV["JULIA_NUM_THREADS"] = "6" (Both passes but tbh it is so much faster with 6), however I think it is better not to have a test directly checking threads as each machine might be different, so instead I updated the CI.yml file to include 2 cases with 1 and 2 threads 

@MaxenceGollier  can you check on your end the tests passes, I added a lock so the issue of the writing to file should be resolved now

dpo commented on src/bmark_solvers.jl:
```diff
@@ -15,11 +20,22 @@ Any keyword argument accepted by `solve_problems`
 #### Return value
 A Dict{Symbol, AbstractExecutionStats} of statistics.
 """
+
```
```suggestion
```

dpo commented on src/bmark_solvers.jl:
```diff
@@ -1,9 +1,14 @@
-export bmark_solvers
+using Base.Threads
 
+export bmark_solvers
```
```suggestion
export bmark_solvers

```

dpo commented on test/test_bmark.jl:
```diff
@@ -11,6 +12,15 @@ function (solver::CallableSolver)(nlp::AbstractNLPModel; kwargs...)
   return GenericExecutionStats(nlp)
 end
 
+function bmark_solvers_signle_thread(solvers::Dict{Symbol, <:Any}, args...; kwargs...)
```
```suggestion
function bmark_solvers_single_thread(solvers::Dict{Symbol, <:Any}, args...; kwargs...)
```

dpo commented on test/test_bmark.jl:
```diff
@@ -65,6 +75,57 @@ function test_bmark()
+      )
+
+      # Run the single-threaded version
+      single_threaded_result = bmark_solvers_signle_thread(solvers, problems)
```
```suggestion
      single_threaded_result = bmark_solvers_single_thread(solvers, problems)
```

dpo commented on test/test_bmark.jl:
```diff
@@ -65,6 +75,57 @@ function test_bmark()
+
+        @test single_threaded_result[key][1, :name] == multithreaded_result[key][1, :name]
+        @test single_threaded_result[key][2, :name] == multithreaded_result[key][2, :name]
+        @test single_threaded_result[key][3, :name] == multithreaded_result[key][3, :name]
```
Could you replace these with loops over the rows of the dataframes?

You could also check the solver-specific info.

farhadrclass replied:
I suggest we actually compare the dataframes directly (or approximate ),

dpo reviewed (changes requested):
Thank you. Almost there.

dpo commented on src/bmark_solvers.jl:
```diff
@@ -16,10 +22,27 @@ Any keyword argument accepted by `solve_problems`
+        # Make a deep copy of the problems for this task so the race condition doesn't occur
+        # local_args = deepcopy.(args)
+        local_args = args
+        local_solver = deepcopy(solver)
```
I'm not sure I understand what this is. At this point, the solver has not yet been instantiated, right? So this is just a copy of a function?!

farhadrclass commented:
@MaxenceGollier  can you check if your example works now ?

MaxenceGollier commented:
> @MaxenceGollier can you check if your example works now ?

Nope, still having the same issue, I think it is more a CUTEst related problem than a problem with your implementation. I will try to see today what is going on here. 
It is weird that it works on your machine, it shouldn't be working on Windows at all...

MaxenceGollier commented:
@farhadrclass Actually, the example shouldn't be working at all on any platform. 
According to @amontoison,  you can not have two instances of the same problem on two different threads for CUTEst problems. I think this is what is causing the issue. We should either add a check on the type of the `AbstractNLPModels` and send a warning/ force the code to use only one thread when the problems are `CUTEstNLPModel` types or add some kind of locks for the problems when the problems are `CUTEstNLPModels`. 

It is really weird that it works on your machine, can you try again ? I just put the code above in a file named `test.jl` in the test folder and added `include("test.jl")` in `runtests.jl`, it then produced the exception. Can you check `Threads.nthreads()` as well when you run this ?

farhadrclass commented on src/bmark_solvers.jl:
```diff
@@ -16,10 +22,22 @@ Any keyword argument accepted by `solve_problems`
+    for (name, solver) in solvers
+        idx += 1
+        names[idx] = name
+        tasks[idx] = Threads.@spawn begin
```
From Julia docs:

> Create a [Task](https://docs.julialang.org/en/v1/base/parallel/#Core.Task) and [schedule](https://docs.julialang.org/en/v1/base/parallel/#Base.schedule) it to run on any available thread in the specified threadpool (:default if unspecified). The task is allocated to a thread once one becomes available. To wait for the task to finish, call [wait](https://docs.julialang.org/en/v1/base/parallel/#Base.wait) on the result of this macro, or call [fetch](https://docs.julialang.org/en/v1/base/parallel/#Base.fetch-Tuple{Task}) to wait and then obtain its return value.

I believe so , as it did on my computer

farhadrclass commented:
> using SolverCore
> using CUTEst
> 
> import SolverCore.dummy_solver
> 
> function test_cutest()
>   problem_names = CUTEst.select(min_con=1, max_con = 3, max_var= 3, only_equ_con=true, only_free_var=true)[1:5]
>   problem_list = (CUTEstModel(name) for name in problem_names)
>   solvers = Dict(:dummy_1 => dummy_solver, :dummy_2 => dummy_solver)
>   stats = bmark_solvers(solvers, problem_list)
> end
> 
> test_cutest()

I am testing with CUTes now, 
I added the unit tests

dpo commented on test/CUTEst_test.jl:
```diff
@@ -0,0 +1,13 @@
+  problem_names = CUTEst.select(min_con=1, max_con = 3, max_var= 3, only_equ_con=true, only_free_var=true)[1:5]
+  problem_list = (CUTEstModel(name) for name in problem_names)
+  solvers = Dict(:dummy_1 => dummy_solver, :dummy_2 => dummy_solver)
+  stats = bmark_solvers(solvers, problem_list)
```
Use the same solvers as before and compare with the single-threaded stats.

farhadrclass replied:
it does not run yet for multi-thread at all,

farhadrclass commented:
> ```julia
> File cannot be deleted
> ```

Thanks @amontoison  I will do that but it is clearly the race condition here as two or more try to write a same file,

farhadrclass commented:
@dpo  I think this is ready
@MaxenceGollier for CUTES, I ran this and it passes
```
using SolverCore
using CUTEst

import SolverCore.dummy_solver

function test_cutest()
  problem_names =
    CUTEst.select(min_con = 1, max_con = 3, max_var = 3, only_equ_con = true, only_free_var = true)[1:5]
  problem_list = (CUTEstModel(name) for name in problem_names)
  solvers = Dict(:dummy_1 => dummy_solver, :dummy_2 => dummy_solver)
  stats = bmark_solvers(solvers, problem_list, use_threads = true)
end

test_cutest()
```
